Performance improvements

SGoetschi opened the pull request:
The main performance issue was our custom extension on how to grab the node hierarchy:
We used an `std::deque` and looped over it to find the right entry. But contrary to expectation the entry set was much larger leading to very poor performance when searching entries.

Additionally, some other performance improvements were found:
* `operator>>` can be quite slow. Moving to `std::strtod` and `std::strtof` yields much faster code (at least on Windows).
* `std::tolower` is comparatively complex (looks at locales etc.) Since we know that it's only used on CityGML tag names, it's ok to only transform the regular ASCII characters which makes it decidedly faster.
* Assembling a string using stringstreams is slow. As we make heavy use of `Attributes::getCityGMLIDAttribute()` this becomes a bottleneck. Moving to `std::string` concatenation and `std::to_string()` yields faster code.

SGoetschi committed:
Replace slow operator >> with faster strto(d|f).

This also fixes a bug with one of our test asset
where not all texture coordinates could be read.

SGoetschi committed:
Replace tolower with simpler and faster version.

This is ok because the CityGML schema only uses
characters in the ASCII space. Therefore we don't
need to take care of locales etc but can just
adapt A-Z to a-z.

SGoetschi committed:
Replace slow stringstream with std::to_string.

SGoetschi committed:
Use unordered_map for IntermediateNode hierarchy.

std::find over the std::deque was very slow for meshes with
large numbers of faces.

SGoetschi committed:
Cleanup: Remove unnecessary  argument.

mistafunk commented on sources/include/parser/parserutils.hpp:
```diff
@@ -46,6 +37,57 @@ namespace citygml {
+        return std::strtod(str, end);
+    }
+
+    template<typename T,
+            std::enable_if_t<std::is_floating_point_v<T>, bool> = true>
```
`enable_if` is technically not necessary, as T is only instantiated for float or double at the moment.

To get this compiled with C++11 (and GCC) you'll need:
```suggestion
    template<typename T,
            typename std::enable_if<std::is_floating_point<T>::value, bool>::type = true>
```

SGoetschi replied:
Thanks for the pointer. I'll remove it. It's not necessary.

mistafunk commented on sources/include/parser/parserutils.hpp:
```diff
@@ -46,6 +37,57 @@ namespace citygml {
+    template<class T> inline T parseNumber(const char* str, char** end);
+
+    template<> inline float parseNumber<float>(const char* str, char** end) {
+        return std::strtof(str, end);
```
`strtof` and co are locale-dependent, are we safe from things like period vs comma here?

SGoetschi replied:
Yes, they are locale dependent. But so was `operator>>`.
I.e. this change doesn't influence the library's behavior.

mistafunk commented on sources/include/parser/parserutils.hpp:
```diff
@@ -61,17 +103,18 @@ namespace citygml {
-        ss << s;
-
-        T v;
         std::vector<T> vec;
```
maybe reserve a typical expected size so the push_back below gets a bit more efficient?

SGoetschi replied:
We could. But I wanted to keep the behavior close to the way it was before - which also didn't reserve anything.

mistafunk commented on sources/include/citygml/citygmlfactory.h:
```diff
@@ -37,7 +37,7 @@ namespace citygml {
         CityModel* createCityModel(const std::string& id);
         CityObject* createCityObject(const std::string& id, CityObject::CityObjectsType type);
-        Geometry* createGeometry(const std::string& id, const CityObject::CityObjectsType& cityObjType = CityObject::CityObjectsType::COT_All, unsigned int lod = 0, std::string srsName = "");
+        Geometry* createGeometry(const std::string& id, const CityObject::CityObjectsType& cityObjType = CityObject::CityObjectsType::COT_All, unsigned int lod = 0, std::string&& srsName = "");
```
why the special rvalue handling of `srsName`? is it that big of a string?

SGoetschi replied:
It's not so much the size of the string but the number of times we call this. We call this once for each face.
This means we can save quite some time avoiding re-instantiating the same string three times in a call hierarchy.
We can also use a `const&` if you'd be happier with that.

mistafunk replied:
more a question of consistency (and least surprise) with the rest of the code - given "number of calls" argument, then the other arguments (e.g. id) should also be rvalues... up to you, I'm not insisting :-)

mistafunk commented on sources/src/citygml/cityobject.cpp:
```diff
@@ -410,34 +407,24 @@ namespace citygml {
     {
-        if (m_NodeStack.find(parentId) == m_NodeStack.end())
-            m_NodeStack.insert(std::make_pair(parentId, std::deque<citygml::IntermediateNode>()));
+        auto const parentIter = m_NodeStack.emplace(parentId, std::unordered_map<std::string, citygml::IntermediateNode>{}).first;
```
Maybe pull out `std::unordered_map<std::string, IntermediateNode>` into a named type?

SGoetschi replied:
This was reworked and probably not necessary anymore.

mistafunk commented on sources/include/citygml/cityobject.h:
```diff
@@ -164,7 +164,7 @@ namespace citygml {
 
         // Keyed on parent ids
-        std::map<std::string, std::deque<IntermediateNode>> m_NodeStack;
+        std::map<std::string, std::unordered_map<std::string, IntermediateNode>> m_NodeStack;
```
"stack" is an unexpected name for two nested maps... maybe there is a better name?

SGoetschi replied:
This triggered a major rework. This map was quite stupid and should have been set up differently anyway.
I created a cleanup commit to deduplicate the two instances of this and another one to rework it.

mistafunk reviewed (changes requested):
Nice work, a few minor things.

A bit sad that we have to fall back to these low-level C `strtof` calls and scattering `char**` left and right... but hey, still limited to C++11.

SGoetschi committed:
Deduplicate node hierarchy tracking.

SGoetschi committed:
Setup hierarchy map from child to parent.

It's inefficient to store it from parent to child. We always need to look through all parents to find the child.

SGoetschi commented:
Needed to force push to clean up the `std::enable_if` in the very first commit. Other than that the commits you already reviewed are untouched - just rebased. The last two commits are new.

mistafunk commented on sources/include/citygml/hierarchyTracker.h:
```diff
@@ -0,0 +1,30 @@
+    {
+    public:
+        void pushIntermediateNode(const IntermediateNode& node, const std::string& parentId);
+        std::string getPreviousParentId(std::string currentParentId, std::shared_ptr<citygml::CityGMLLogger> logger) const;
```
It has been like that before, but is `currentParentId` intentionally passed by value to facilitate RVO? I think the majority of cases return a new string, so pass by ref could be justified.

SGoetschi replied:
I didn't really think about this. But let's keep the old behavior to avoid any surprises.

mistafunk commented on sources/src/citygml/hierarchyTracker.cpp:
```diff
@@ -28,48 +28,32 @@ namespace citygml {
 
-        // if replacing the root then root elements are moved to become children of the new element
+        // if replacing the root then the root element is moved to become a child of the new element
         if (parentId == "root")
```
literal `"root"` appears a few times - could be extracted

SGoetschi replied:
Yeah. Next time.

mistafunk reviewed (approved):
Nice, nothing major. See inline comments.

SGoetschi commented:
Removed the `std::string&&` commit and force-pushed again. That's why the last two commits appear as new but they are exactly the same as before.
